cranelift/codegen/src/egraph/stores.rs

```diff
@@ -0,0 +1,270 @@
+    /// Last-store instruction (or none) for a given load.
+    load_mem_state: FxHashMap<Inst, MemoryState>,
+}
+
+impl AliasAnalysis {
+    /// Perform an alias analysis pass.
+    pub fn new(func: &Function, cfg: &ControlFlowGraph) -> AliasAnalysis {
+        log::trace!("alias analysis: input is:\n{:?}", func);
+        let mut analysis = AliasAnalysis {
+            load_mem_state: FxHashMap::default(),
+        };
+
+        let block_input = analysis.compute_block_input_states(func, cfg);
+        analysis.compute_load_last_stores(func, block_input);
+        analysis
+    }
+
+    #[inline(never)]
+    fn compute_block_input_states(
+        &mut self,
+        func: &Function,
+        cfg: &ControlFlowGraph,
+    ) -> SecondaryMap<Block, LastStores> {
+        let mut block_input = SecondaryMap::with_default(LastStores::default());
+        block_input.resize(func.dfg.num_blocks());
+        let mut queue: SmallVec<[Block; 8]> = smallvec![];
+        let mut queue_set = FxHashSet::default();
+        let entry = func.layout.entry_block().unwrap();
+        queue.push(entry);
+        queue_set.insert(entry);
+        block_input[entry] = LastStores::entry();
+
+        while let Some(block) = queue.pop() {
+            queue_set.remove(&block);
+            let mut state = block_input[block].clone();
+
+            log::trace!(
+                "alias analysis: input to block{} is {:?}",
+                block.index(),
+                state
+            );
+
+            for inst in func.layout.block_insts(block) {
+                state.update(func, inst);
+                log::trace!("after inst{}: state is {:?}", inst.index(), state);
+            }
+
+            for succ in cfg.succ_iter(block) {
+                let succ_first_inst = func.layout.block_insts(succ).into_iter().next().unwrap();
+                let succ_state = &mut block_input[succ];
+                let old = succ_state.clone();
+                succ_state.meet_from(&state, succ_first_inst);
+                let updated = *succ_state != old;
+
+                if updated && queue_set.insert(succ) {
+                    queue.push(succ);
+                }
+            }
+        }
+
+        block_input
+    }
+
+    #[inline(never)]
+    fn compute_load_last_stores(
+        &mut self,
+        func: &Function,
+        block_input: SecondaryMap<Block, LastStores>,
+    ) {
+        for block in func.layout.blocks() {
+            let mut state = block_input[block].clone();
+
+            for inst in func.layout.block_insts(block) {
+                log::trace!(
+                    "alias analysis: scanning at inst{} with state {:?} ({:?})",
+                    inst.index(),
+                    state,
+                    func.dfg[inst],
+                );
+
+                let opcode = func.dfg[inst].opcode();
+                if opcode == Opcode::Load {
+                    let mem_state = state.get_load_input_state(func, inst);
```

Comment:
> I'd like a comment here on why this is currently limited to the Load opcode specifically, rather than any instruction with that format. I assume it's something about store-to-load forwarding requiring that the load and store access the same memory width, and also have the same Value type, so that the value passed to the store can be substituted for the load. (I imagine future work could include replacing widening loads with sign- or zero-extend instructions given a suitable narrow store.) But it's not obvious to me how this interacts with redundant-load elimination.

Yep, that's exactly it. I've added a comment to that effect.

The interaction with RLE is that the RLE pass performs an action only if this analysis provides a "memory state" for an instruction, so if we withhold that state, then the analysis is effectively saying "this is not an analyzable load".

> [removing redundant match, and `for_flags` update]

Awesome, I like this much better -- done!

(I think this is a Lens in the Haskell sense, or at least a Functor, but I'm gonna look away from the abyss before it swallows me here)